### PR TITLE
Add support for associating an OIDC provider with an EKS cluster

### LINKED ISSUE #610: add oidc identity provider to eks
### What problem are you facing?
Currently the only way to login to an eks cluster which was provisioned by `provider-aws` is via IAM user.
Recently AWS added the [feature](https://aws.amazon.com/de/blogs/containers/introducing-oidc-identity-provider-authentication-amazon-eks/) to login to an eks cluster via oidc, but it's not yet supported by `provider-aws`

### How could Crossplane help solve your problem?
The eks resource in `provider-aws` should be updated with the `oidc` configuration options: https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/eks@v1.2.0/types#OidcIdentityProviderConfigRequest

Depends on https://github.com/crossplane/provider-aws/issues/383

### PR DESCRIPTION
**Pull request by goober**
### Description of your changes
This pull request aims to add support for associating an OpenID Connect Provider to an EKS cluster.

It will add a new resource `IdentityProviderConfig` that can be associated to an EKS cluster.

Fixes #610 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

`oidc.yaml`
```yaml
apiVersion: eks.aws.crossplane.io/v1alpha1
kind: IdentityProviderConfig
metadata:
  name: azure-ad
  namespace: default
spec:
  forProvider:
    region: eu-north-1
    clusterName: crossplane-cluster
    oidc:
      clientId: spn:<omitted-azure-ad-clientId>
      issuerUrl: https://sts.windows.net/<omitted-azure-ad-tenant-id>/
      usernameClaim: "upn"
      usernamePrefix: "aad:"
    tags:
      my-tag: value1
```

```bash
$ kubectl apply -f oidc.yaml
identityproviderconfig.eks.aws.crossplane.io/azure-ad created

$ kubectl get identityproviderconfigs
NAME       READY   SYNCED   CLUSTER              AGE
azure-ad   False   True     crossplane-cluster   14s

# wait about 30 minutes until the identityprovider has been attached to the cluster

$ kubectl describe identityproviderconfig azure-ad

Name:         azure-ad
Namespace:
Labels:       <none>
Annotations:  crossplane.io/external-create-pending: 2021-10-25T15:29:47Z
              crossplane.io/external-create-succeeded: 2021-10-25T15:29:49Z
              crossplane.io/external-name: azure-ad
API Version:  eks.aws.crossplane.io/v1alpha1
Kind:         IdentityProviderConfig
Metadata:
  Creation Timestamp:  2021-10-25T15:29:47Z
  Finalizers:
    finalizer.managedresource.crossplane.io
  Generation:  2
  Managed Fields:
    API Version:  eks.aws.crossplane.io/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:deletionPolicy:
        f:forProvider:
          .:
          f:clusterName:
          f:oidc:
            .:
            f:clientId:
            f:issuerUrl:
            f:usernameClaim:
            f:usernamePrefix:
          f:region:
          f:tags:
            .:
            f:my-tag:
        f:providerConfigRef:
          .:
          f:name:
    Manager:      kubectl-client-side-apply
    Operation:    Update
    Time:         2021-10-25T15:29:47Z
    API Version:  eks.aws.crossplane.io/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          f:crossplane.io/external-create-pending:
          f:crossplane.io/external-create-succeeded:
          f:crossplane.io/external-name:
        f:finalizers:
          .:
          v:"finalizer.managedresource.crossplane.io":
      f:spec:
        f:forProvider:
          f:tags:
            f:crossplane-kind:
            f:crossplane-name:
            f:crossplane-providerconfig:
      f:status:
        .:
        f:atProvider:
          .:
          f:identityProviderConfigArn:
          f:status:
        f:conditions:
    Manager:         ___go_build_github_com_crossplane_provider_aws_cmd_provider
    Operation:       Update
    Time:            2021-10-25T15:29:49Z
  Resource Version:  23231
  UID:               e2a44b07-753f-4f0c-8b1e-8418a65b14be
Spec:
  Deletion Policy:  Delete
  For Provider:
    Cluster Name:  crossplane-cluster
    Oidc:
      Client Id:        spn:<omitted-client-id>
      Issuer URL:       https://sts.windows.net/<omitted-tenant-id>/
      Username Claim:   upn
      Username Prefix:  aad:
    Region:             eu-north-1
    Tags:
      Crossplane - Kind:            identityproviderconfig.eks.aws.crossplane.io
      Crossplane - Name:            azure-ad
      Crossplane - Providerconfig:  default
      My-tag:                       value1
  Provider Config Ref:
    Name:  default
Status:
  At Provider:
    Identity Provider Config Arn:  arn:aws:eks:eu-north-1:<omitted-accountId>:identityproviderconfig/crossplane-cluster/oidc/azure-ad/<omitted>
    Status:                        ACTIVE
  Conditions:
    Last Transition Time:  2021-10-25T16:02:01Z
    Reason:                Available
    Status:                True
    Type:                  Ready
    Last Transition Time:  2021-10-25T15:29:49Z
    Reason:                ReconcileSuccess
    Status:                True
    Type:                  Synced
Events:
  Type    Reason                   Age   From                            Message
  ----    ------                   ----  ----                            -------
  Normal  CreatedExternalResource  32m   managed/identityproviderconfig  Successfully requested creation of external resource

```

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by haarchri** [comment]
@goober can you resolve check-diff error? think rerun make generate will solve it

**Comment by goober** [comment]
Thanks @haarchri I have maybe fixed it now. The problem is that when I am running the command locally the go generate will replace all 
`// +build !ignore_autogenerated` with `//go:build !ignore_autogenerated` so it will always fail locally

**Comment by haarchri** [comment]
@goober  sounds like you using locally not the same go-version then provider-aws https://github.com/crossplane/provider-aws/blob/master/go.mod#L3

**Comment by goober** [comment]
Thanks @haarchri that was it!

File: apis/eks/v1alpha1/identityproviderconfig_references.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
+/*
+Copyright 2020 The Crossplane Authors.
```
Comment (haarchri) [inline apis/eks/v1alpha1/identityproviderconfig_references.go:2]: small nitpick: change to 2021 in all files

**Review by haarchri — CHANGES_REQUESTED** [review]
we added in this PR no example because we have no real data for input ? - please check 2 small nitpicks - in general the resource is working

File: pkg/controller/eks/identityproviderconfig/controller.go [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+	}
+
+	current := cr.Spec.ForProvider.DeepCopy()
+	eks.LateInitializeIdentityProviderConfig(&cr.Spec.ForProvider, rsp.IdentityProviderConfig)
+
+	// Generate observation
+	cr.Status.AtProvider = eks.GenerateIdentityProviderConfigObservation(rsp.IdentityProviderConfig)
+
+	// Any of the statuses we don't explicitly address should be considered as
+	// the node group being unavailable.
```
Comment (haarchri) [inline pkg/controller/eks/identityproviderconfig/controller.go:121]: Really the nodeGroup? think during apply/delete nodeGroup is available - so Copy/Paste NitPick?

File: apis/eks/v1alpha1/identityproviderconfig_references.go (hunk 1 above)
Comment (goober) [reply apis/eks/v1alpha1/identityproviderconfig_references.go]: Fixed

File: pkg/controller/eks/identityproviderconfig/controller.go (hunk 2 above)
Comment (goober) [reply pkg/controller/eks/identityproviderconfig/controller.go]: Fixed

**Review by haarchri — APPROVED** [review]

**Comment by haarchri** [comment]
LGTM
Ressource Create & Delete is working overall code looks good - thanks for your Contribution

**Pull request merged**